resolver/resolver.go

```diff
@@ -467,28 +472,29 @@ func truncate(m *dns.Msg, udp bool) *dns.Msg {
 
 	furtherTruncation := m.Len() > max
 	m.Truncated = m.Truncated || furtherTruncation
-
+	if !setTruncateBit {
+		// Clear the Truncate bit regardless of whether this message used
+		// to have it set set or whether we are going to truncate it here.
+		m.Truncated = false
+	}
 	if !furtherTruncation {
 		return m
 	}
-
 	m.Extra = nil // Drop all extra records first
 	if m.Len() < max {
+		// Dropping the extra records shrunk the message down sufficiently.
 		return m
 	}
 	answers := m.Answer[:]
-	left, right := 0, len(m.Answer)
-	for {
-		if left == right {
-			break
-		}
-		mid := (left + right) / 2
-		m.Answer = answers[:mid]
-		if m.Len() < max {
-			left = mid + 1
-			continue
-		}
-		right = mid
+	search := func(n int) bool {
+		m.Answer = answers[:n]
+		return m.Len() > max
+	}
+	limit := sort.Search(len(answers), search)
+	if limit > 0 {
+		m.Answer = answers[:limit-1]
```

Comment:
No - `sort.Search` returns the first value of `n` for which the bool check `m.Len() > max` is true. Looking at the search function it is clear that this is the value `n` such that `m.Answer[:n]` results in `m.Len() > max`.

Perhaps a better approach would be to search in the reverse direction, then we'd return the first `n` for which `m.Len() <= max`. I'll see if I can make this code simpler, but it certainly is correct as is.